### ISSUE TITLE
100% CPU Usage

### ISSUE THREAD

**Issue by OD-C1** [cloudflare/cloudflared#302]
Running version 2021.1.5 on Odroid C1 with Ubuntu 20.04 together with PiHole and Wireguard. After many days all working  flawless there can be a connection lost to the internet caused by the ISP. Shortly after that, the CPU usage for cloudflared service goes to 100%. Device need either to be rebooted or cloudflared service restarted. Problem have been there on older versions.

**Comment by adamchalmers** [comment]
Hi, thanks for the bug report, we'll look into it. Do you have logs from the time this happened?

**Comment by OD-C1** [comment]
> Hi, thanks for the bug report, we'll look into it. Do you have logs from the time this happened? Your config file and CLI invocation would also be helpful.

Hi Adam, thanks for your reply. Please see attached.
[Config.txt](https://github.com/cloudflare/cloudflared/files/5906085/Config.txt)
[Logs.txt](https://github.com/cloudflare/cloudflared/files/5906087/Logs.txt)
[cloudflared.service.txt](https://github.com/cloudflare/cloudflared/files/5906088/cloudflared.service.txt)

**Comment by AlexKF98** [comment]
I can confirm this issue! Internet goes out because of ISP, when the internet returns, there's a memory leak, CPU usage 100% and overheating. I'm using cloudflared to have DNS over HTTPS with my Pi-Hole.

This is what Pi-Hole looks like with normal usage:

![Captura de tela de 2021-02-02 14-19-26](https://user-images.githubusercontent.com/12222211/106637709-cd910b00-6561-11eb-9c79-0f1f78586a6f.png)

After internet goes out because of ISP and then returns:

![Captura de tela de 2021-02-02 00-38-59](https://user-images.githubusercontent.com/12222211/106637793-e7cae900-6561-11eb-8721-1130654412c9.png)

htop with memory leak, CPU load and overheating:

![Captura de tela de 2021-02-02 00-38-12](https://user-images.githubusercontent.com/12222211/106638048-2e204800-6562-11eb-852d-6d0e98b856c4.png)

**Comment by xxKeith20xx** [comment]
I'm not able to replicate on m pi3b+ over ethernet. I'm running the following:

raspian 10
kernel 5.4.79
cloudflared version 2021.1.5 (built 2021-01-18-1220 UTC)
Pi-hole version is v5.2.4 (Latest: v5.2.4)
AdminLTE version is v5.3.2 (Latest: v5.3.2)
FTL version is v5.6 (Latest: v5.6)

```
root # systemctl status cloudflared
● cloudflared.service - Argo Tunnel
   Loaded: loaded (/etc/systemd/system/cloudflared.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2021-02-03 18:08:02 CST; 1min 32s ago
 Main PID: 29148 (cloudflared)
    Tasks: 9 (limit: 2182)
   CGroup: /system.slice/cloudflared.service
           └─29148 /usr/local/bin/cloudflared --config /etc/cloudflared/config.yml --no-autoupdate --origincert /etc/cloudflared/cert.pem

Feb 03 18:08:01 pihole systemd[1]: Starting Argo Tunnel...
Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Version 2021.1.5
Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF GOOS: linux, GOVersion: go1.15.6, GoArch: arm
Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Settings: map[config:/etc/cloudflared/config.yml no-autoupdate:true origincert:/etc/cloudflared/cert.pem proxy-dns:true
Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Adding DNS upstream url=[omitted]/dns-query
Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Starting DNS over HTTPS proxy server address=dns://localhost:5053
Feb 03 18:08:02 pihole systemd[1]: Started Argo Tunnel.
```

**Comment by AlexKF98** [comment]
> I'm not able to replicate on my pi3b+ over ethernet after unplugging connection briefly to simulate.
> 
> I'm running the following:
> 
> raspian 10
> kernel 5.4.79
> cloudflared version 2021.1.5 (built 2021-01-18-1220 UTC)
> Pi-hole version is v5.2.4 (Latest: v5.2.4)
> AdminLTE version is v5.3.2 (Latest: v5.3.2)
> FTL version is v5.6 (Latest: v5.6)
> 
> I restarted the service and then tested:
> 
> ```
> root # systemctl status cloudflared
> ● cloudflared.service - Argo Tunnel
>    Loaded: loaded (/etc/systemd/system/cloudflared.service; enabled; vendor preset: enabled)
>    Active: active (running) since Wed 2021-02-03 18:08:02 CST; 1min 32s ago
>  Main PID: 29148 (cloudflared)
>     Tasks: 9 (limit: 2182)
>    CGroup: /system.slice/cloudflared.service
>            └─29148 /usr/local/bin/cloudflared --config /etc/cloudflared/config.yml --no-autoupdate --origincert /etc/cloudflared/cert.pem
> 
> Feb 03 18:08:01 pihole systemd[1]: Starting Argo Tunnel...
> Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Version 2021.1.5
> Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF GOOS: linux, GOVersion: go1.15.6, GoArch: arm
> Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Settings: map[config:/etc/cloudflared/config.yml no-autoupdate:true origincert:/etc/cloudflared/cert.pem proxy-dns:true
> Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Adding DNS upstream url=[omitted]/dns-query
> Feb 03 18:08:02 pihole cloudflared[29148]: 6:08PM INF Starting DNS over HTTPS proxy server address=dns://localhost:5053
> Feb 03 18:08:02 pihole systemd[1]: Started Argo Tunnel.
> ```
> 
> ```
> root # cat /proc/loadavg
> 0.12 0.19 0.18 1/157 2494
> ```
> 
> ```
> SYSTEM ========================================================================
>  Uptime:   35 days, 9 hours, 38 minutes            Memory:   [■■········] 16.8%
>  CPU Temp: 122.8°F   CPU Load: 0.21, 0.21, 0.18    CPU Load: [■·········] 5.2%  
> ```

I'm using a separate modem and router. Pi-Hole is linked to the router through an ethernet cable. I think this issue occurs when the modem can't connect to the internet, but the router continues to operate nevertheless (which means the network is still on but can't access external sites). Then CPU usage goes to 100%, memory usage increases and Raspberry Pi starts to overheat.

**Comment by xxKeith20xx** [comment]
@AlexKF98 how long does it usually take to climb that high? I tried again by disabling ethernet port from the router and letting it stay like that for a few minutes but didn't see any jump in CPU still. Maybe I'd need to actually disconnect the cable connection to test?

**Comment by AlexKF98** [comment]
> @AlexKF98 how long does it usually take to climb that high? I tried again by disabling ethernet port from the router and letting it stay like that for a few minutes but didn't see any jump in CPU still. Maybe I'd need to actually disconnect the cable connection to test?

This is how I’m able to reproduce it, I’ve noticed the problem appears after internet connection is resumed.
1. Turn off the internet modem. Don’t touch the router or the Pi-Hole, keep the router on and leave the Pi-Hole connected to it through an Ethernet cable. So local IPs can be accessed, but not the internet.
2. Turn the modem on again, internet connection returns.
3. After internet connection is resumed, CPU usage jumps to 100% and Raspberry Pi quickly overheats

**Comment by xxKeith20xx** [comment]
I'm still not able to replicate but I'm still having to simulate it a bit.

So with my setup, I have the cable modem connected to a Mikrotik hAP ac² that does my wifi/ethernet. I logged into the Mikrotik interface and deactivated just the port that connects to the cable modem, thus allowing my local network to function normally but getting no access to the internet. I did see unbound/cloudflared spike to 2% or 4% CPU for a few seconds while it couldn't reach out, but it never went above that. I let it run for about 5 minutes before turning it back on.

It may be worth seeing if you can replicate without an actual internet outage and see if you can capture any log output? One thing I noticed is that my pihole log went a bit nuts when it couldn't reach the internet. I'm wondering if this is more related to logging/pihole now? I wonder if there is a I/O issue during that CPU spike as well? Especially since it's running off a SDCARD (presumably - mine is).

If you happened to have sysstat running, those resource logs may offer some clues too. May be worth posting to https://discourse.pi-hole.net as well. Something performance specific to the setup, like the situation in https://discourse.pi-hole.net/t/pihole-ftl-using-all-my-cpu-and-breaks-all-internet-connectivity/15672 might be relevant. For example, in the logs @OD-C1 provided, the frequency of `Feb 01 12:27:46 odroid cloudflared[551]: 12:27PM ERR failed to connect to an HTTPS backend "https://1.1.1.1/dns-query" error="failed to perform an HTTPS request: Post \"https://1.1.1.1/dns-query\": context deadline exceeded"` is pretty high and I'd be curious to see if there were I/O spikes there.

**Comment by adamchalmers** [comment]
This should be fixed on master ([PR](https://github.com/cloudflare/cloudflared/pull/290), [commit](https://github.com/cloudflare/cloudflared/commit/d7c4a89106268815760ed0efc7c08bf506d58517)) and will be available in the next release 2021.2.3.
